core/tls.go

```diff
@@ -0,0 +1,80 @@
+package core
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"os"
+
+	"chain/errors"
+)
+
+var ErrNoTLS = errors.New("no TLS configuration available")
+
+// TLSConfig returns a TLS config suitable for use
+// as a Chain Core client and server.
+// It reads a PEM-encoded X.509 certificate and private key
+// from certFile and keyFile.
+// If rootCAs is given, it reads a list of trusted root CA certs
+// from the filesystem;
+// otherwise it uses the system cert pool.
+//
+// For compatibility, it attempts to read the cert and key
+// from the environment if certFile and keyFile both
+// do not exist in the filesystem.
+//   TLSCRT=[PEM-encoded X.509 certificate]
+//   TLSKEY=[PEM-encoded X.509 private key]
+//
+// If certFile and keyFile do not exist or are empty
+// and the environment vars are both unset,
+// TLSConfig returns ErrNoTLS.
+func TLSConfig(certFile, keyFile, rootCAs string) (*tls.Config, error) {
+	config := &tls.Config{
+		// This is the default set of protocols for package http.
+		// ListenAndServeTLS and Transport set this automatically,
+		// but since we're supplying our own TLS config,
+		// we have to set it here.
+		NextProtos: []string{"http/1.1", "h2"},
```

Comment:
So this part of the HTTP package is really hard to follow and not well documented. But that slice of strings is only used to initialized this map if the map is unset; our own assignment in cmd/cored will take precedence even if we set TLSConfig on the Server.

See https://github.com/golang/go/blob/48163968b2927247213fca7a6f4678d3c93855dc/src/net/http/server.go#L2946-L2948.